First commit prometheus middleware.

enxebre opened the pull request:
I started to look at this so I wanted to get familiar with the source code.
This is a trivial prometheus middleware based on https://github.com/zbindenren/negroni-prometheus

It could be an starting point for https://github.com/containous/traefik/issues/839
I'm happy to continue working on this if there's interest on this feature.

As we've agreed on some points and this is taking shape I'll add a progress tracker here:
- [x] Discuss design
- [x] Define Metrics middleware and Interface
- [x] Create implementation for Prometheus
- [x] Agree on basic labels 
- [x] Ability to choose exporter at run time (flag)
```--web.metrics.prometheus```
- [x] Support configurable params for exporters at runtime
```--web.metrics.prometheus.Buckets="100,200"```
- [x] Update Docs

emilevauge commented:
ping @SantoDE :)

SantoDE commented:
Hey @enxebre ,

thank's a lot for your contribution :-)

You're idea is in the right direction and a good starting point, but I'm looking for a slightly different approach to not "vendor lock in" on just prometheus. If you're interested in going on further here, let us have a short talk on Slack and / or drop me an email <ich@manuel-laufenberg.de>. I guess, together we can move that further fast :)

/cc: @containous/traefik 

Regards,
Santo

ralphtheninja commented:
It would be awesome to have some kind of support for Prometheus. Looking forward to what you can come up with! :)

bamarni commented:
There seems to be a nice generic package for that purpose : https://github.com/rcrowley/go-metrics

Looking forward to this too 👍

enxebre commented:
cool, lets discuss on slack then. @SantoDE

enxebre commented:
Potential solutions for a more generic approach could be:
- https://github.com/go-kit/kit/tree/master/metrics which already includes support for prometheus
- https://github.com/rcrowley/go-metrics + define a custom interface + specific implementations (e.g https://github.com/deathowl/go-metrics-prometheus or https://github.com/armon/go-metrics/blob/master/prometheus/prometheus.go)

enxebre commented:
Added a second commit with a more vendor neutral approach using github.com/go-kit/kit/metrics

emilevauge commented on glide.yaml:
```diff
@@ -101,7 +101,7 @@ import:
   version: ^v1.5.0
 - package: github.com/gogo/protobuf
-  version: 0.3
+  version: "0.3"
```
why any `"` here ?

enxebre replied:
I think it was just glide adding that

bamarni commented on middlewares/prometheus.go:
```diff
@@ -0,0 +1,64 @@
+)
+
+var (
+	dflBuckets = []float64{300, 1200, 5000}
```
300ms seems high already, as those are cumulative how about adding some lower ones too?

SantoDE replied:
Maybe 100?

bamarni commented on middlewares/metrics.go:
```diff
@@ -0,0 +1,39 @@
+	start := time.Now()
+	prw := &responseRecorder{rw, http.StatusOK}
+	next(prw , r)
+	labels := []string{ "code", strconv.Itoa(prw.StatusCode()), "method", r.Method, "path", r.URL.Path }
```
I'm not sure the request path is suitable as a label for a generic metrics handler, because it has an unbounded cardinality.

SantoDE replied:
We should probably use something else then the path. Sadly, I can't think of something yet. Will update once I have an idea :-)

bamarni commented on middlewares/prometheus.go:
```diff
@@ -0,0 +1,64 @@
+		[]string{"code", "method", "path"},
+	)
+
+	if len(buckets) == 0 {
```
At the moment this check seems unneeded as the buckets are not configurable.

SantoDE commented on server.go:
```diff
@@ -680,6 +682,8 @@ func (server *Server) loadConfig(configurations configs, globalConfiguration Glo
 						}
 
 						var negroni = negroni.New()
+						metricsMiddlewareBackend := middlewares.NewMetricsWrapper(middlewares.NewPrometheus(frontend.Backend))
```
We should make it configurable (for the feature), which specific "exporter" to use. WDYT @enxebre?

enxebre replied:
yup, I agree

SantoDE commented on glide.yaml:
```diff
@@ -113,4 +113,10 @@ import:
-- package: github.com/mvdan/xurls
\ No newline at end of file
+- package: github.com/mvdan/xurls
+- package: github.com/prometheus/client_golang
```
Do we need the client, if we're going with go-kit? Shouldn't go-kit bring it's dependencies?

enxebre replied:
You're right

SantoDE commented:
Hey @enxebre,

thank's for your contribution 👍 

I'm okay with using go-kit as it's supporting labels (over e.g. go-metrics, which doesn't). I also like the approach of having a metrics middleware which then uses the specific "exporter". I added some feedback to the Review.

enxebre commented:
Added a new commit with updated tests, docs and runtime config for monitoring.
Some explicit logic for the promhttp.handler on web.go and for instantiating the monitoring implementation on server.go is intentionally left there. This can be hidden in the metricsWrapper in the future once we have a least two implementations so we can see the right abstractions. https://github.com/containous/traefik/pull/1022/commits/2fe8c82398eeb3c355e3220fa9f4e1f42e1823d9

SantoDE commented:
Hey @enxebre,

that looks good 👍 When do you will have time to fix the conficts?

enxebre commented:
Conflicts fixed and built green now

SantoDE commented:
Hey @enxebre ,

Thanks again! Really :)

LGTM 👼 

Could you please squash and rebase your branch? Thanks! :-)

bamarni commented:
How about the path label, isn't the backend label enough? My concern with it is that it might create a lot of time series, if for instance backends contain dynamic placeholders such as order_id or user_id, which is very common.

IMO a label based on the request path is more suited for backends which have a well-known and limited set of paths, thus should be registered explicitly at application level as we don't know about it at the proxy level.

SantoDE commented:
@enxebre Yeah, I would probably keep out the path labels (as mentioned in the review). I don't see any real advantage just yet. A label per backend should be fine :)

enxebre commented:
@bamarni @SantoDE removed path label, thanks for the comments

SantoDE commented:
@enxebre You're welcome! Now it needs a rebase again ;-)

enxebre commented on configuration.go:
```diff
@@ -327,6 +327,9 @@ func NewTraefikDefaultPointersConfiguration() *TraefikConfiguration {
 		RecentErrors: 10,
 	}
+	defaultWeb.Metrics = &types.Metrics{
+		System: "prometheus",
```
I just wanted to avoid exporter as It's a bit confusing with https://prometheus.io/docs/instrumenting/exporters/ but It's fine, Exporter probably better than System anyway

enxebre commented:
Hey @emilevauge, as I mentioned above once we add a second "Metrics Exporter implementation" (If this uses a push approach e.g statsd) will be easy to realise the best way to abstract the methods so both pullers and pushers can potentially implement the same interface.
Just changed System for Exporter and removed the glide thing

emilevauge commented:
@enxebre thanks a lot :)
>Integration test for Eureka TestSimpleConfiguration is failing, don't think it relates to this PR. Any idea?

Yeah some tests fail sometimes... We have to clean this up ;)

While reading your code again, I was thinking that in order to stay homogeneous with the rest of the code, we could rewrite the config structs like this:

```
// Metrics provides options to expose and send Traefik metrics to different third party monitoring systems
type Metrics struct {
	Prometheus *Prometheus `description:"Prometheus metrics exporter type"`
}

type Prometheus struct {
}
```

We could enable Prometheus using the boolean arg `--web.metrics.prometheus`.
Besides, this would allow specific configuration by third party monitoring system.

WDYT?
ping @SantoDE

SantoDE commented:
Oh, I would like that idea! That makes totally sense :) 👍

enxebre commented:
I like it. How about something like https://github.com/containous/traefik/pull/1022/commits/4477b339c231f0329de11e97f13fe1a728da39c3 What's the best way to pass [] from the command line? Would it be possible for Prometheus.Buckets to be a [] rather than a string?

enxebre commented:
Never mind, just found the way :) https://github.com/containous/traefik/pull/1022/commits/3ce9d690f91039425d3b7bd2e9ed077b8fba4907

enxebre commented:
Cool, green build now 🙌

SantoDE commented:
Merged 👍 👼 

Thanks a lot @enxebre

SantoDE commented:
Merged 👍 👼 

Thanks a lot @enxebre

ralphtheninja commented on docs/toml.md:
````diff
@@ -715,6 +715,11 @@ $ curl -s "http://localhost:8080/api" | jq .
+
+```bash
+$ traefik --web.metrics.prometheus --web.metrics.prometheus.buckets="100,300"
+```
````
Is this all it takes? No steps needed to configure prometheus?

enxebre replied:
@ralphtheninja this is how Traefik exposes metrics to Prometheus. As a separate thing you must run Prometheus and configure it to scrape the Traefik endpoint . It be nice to have a Kubernetes Helm chart example for Traefik + Prometheus + Grafana dashboard showing the whole thing

ralphtheninja replied:
Ok. Thanks for info!

brian-brazil commented on middlewares/prometheus.go:
```diff
@@ -0,0 +1,65 @@
+
+const (
+	reqsName    = "requests_total"
+	latencyName = "request_duration_milliseconds"
```
The Prometheus convention is to use seconds.

brian-brazil commented on middlewares/prometheus.go:
```diff
@@ -0,0 +1,65 @@
+// NewPrometheus returns a new prometheus Metrics implementation.
+func NewPrometheus(name string, config *types.Prometheus) *Prometheus {
+	var m Prometheus
+	m.reqsCounter = prometheus.NewCounterFrom(
```
This is already covered by the latency histogram's _count timeseries.

It's recommend not to split out latency results by success/failure/status code as that's very commonly misinterpreted. What I'd suggest is removing the code label from the latency metric.

It's also unwise on cardinality grounds to have too many labels on a histogram, as it already has one itself.

brian-brazil commented:
> You're idea is in the right direction and a good starting point, but I'm looking for a slightly different approach to not "vendor lock in" on just prometheus.

The way we handle this in Prometheus is that the client libraries aren't tied to Prometheus, so you can for example instrument with Prometheus and output to Graphite without ever running a Prometheus server. Thus we recommend to use the standard Prometheus client libraries where possible.

SantoDE commented:
Hey @brian-brazil,

thank's a lot for your valuable feedback :-) Do you have any more? Shall we have a talk on Slack? Or are you even interested in putting up a PR on your? 

This one is already merged so we need to put up another one anyways.

Cheers

brian-brazil commented:
I saw it mentioned it on twitter, so I put in a few comments on common issues around instrumentation. 

From another look, the only other comment I'd have would be to prefix your metrics with traefik_ or whatever else makes sense for you. In theory (and it's very much a theory) metric names should be globally unique, and you want a user to have an idea of what sort of request it is and where it's being measured from.

enxebre commented:
Hey @brian-brazil thanks for the feedback!
I'll create a follow up PR to address the comments about instrumentation
